assets/js/googlesitekit/datastore/modules/modules.js

```diff
@@ -0,0 +1,434 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from './constants';
+
+const { createRegistrySelector } = Data;
+
+// Actions
+const START_FETCH_SET_MODULE_STATUS = 'START_FETCH_SET_MODULE_STATUS';
+const FETCH_SET_MODULE_STATUS = 'FETCH_SET_MODULE_STATUS';
+const FINISH_FETCH_SET_MODULE_STATUS = 'FINISH_FETCH_SET_MODULE_STATUS';
+const CATCH_FETCH_SET_MODULE_STATUS = 'CATCH_FETCH_SET_MODULE_STATUS';
+const RECEIVE_SET_MODULE_STATUS = 'RECEIVE_SET_MODULE_STATUS';
+const START_FETCH_MODULES = 'START_FETCH_MODULES';
+const FETCH_MODULES = 'FETCH_MODULES';
+const FINISH_FETCH_MODULES = 'FINISH_FETCH_MODULES';
+const CATCH_FETCH_MODULES = 'CATCH_FETCH_MODULES';
+const RECEIVE_MODULES = 'RECEIVE_MODULES';
+
+export const INITIAL_STATE = {
+	modules: undefined,
+	isFetchingSetModuleStatus: {},
+	isFetchingModules: false,
+};
+
+export const actions = {
+	/**
+	 * Activate a module on the server.
+	 *
+	 * Activate a module (based on the slug provided).
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {string} slug Slug of the module to activate.
+	 * @return {Object}      Object with {response, error}
+	 */
+	*activateModule( slug ) {
+		const { response, error } = yield actions.setModuleStatus( slug, true );
+		return { response, error };
+	},
+
+	/**
+	 * Dectivate a module on the server.
+	 *
+	 * Dectivate a module (based on the slug provided).
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {string} slug Slug of the module to activate.
+	 * @return {Object}      Object with {response, error}
+	 */
+	*deactivateModule( slug ) {
+		const { response, error } = yield actions.setModuleStatus( slug, false );
+		return { response, error };
+	},
+
+	/**
+	 * (De)activate a module on the server.
+	 *
+	 * POSTs to the `core/modules/activation` endpoint to set the `active` status
+	 * supplied for the give `slug`.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param  {string}  slug   Slug of the module to activate/deactivate.
+	 * @param  {boolean} active `true` to activate; `false` to deactivate.
+	 * @return {Object}         Object with {response, error}
+	 */
+	*setModuleStatus( slug, active ) {
+		invariant( slug, 'slug is required.' );
+		invariant( active !== undefined, 'active is required.' );
+
+		let response, error;
+
+		yield {
+			payload: { active, slug },
+			type: START_FETCH_SET_MODULE_STATUS,
+		};
+
+		try {
+			response = yield {
+				payload: { active, slug },
+				type: FETCH_SET_MODULE_STATUS,
+			};
+
+			if ( response.success === true ) {
+				// Fetch (or re-fetch) all modules, with their updated status.
+				yield actions.fetchModules();
```

Comment:
Nice! 👍 

I guess this is mainly to account for dependency modules? Theoretically otherwise we could just modify the one module activated/deactivated in the store directly, but I see making the extra request for fresh data is a more bulletproof approach.

Reply:
Yeah, I was thinking anything that's got dependencies/interconnected-ness would benefit from this. We can already use this, but third-party modules might especially benefit from this.